apps/api/src/app/shared/framework/idempotency.interceptor.ts

```diff
@@ -0,0 +1,235 @@
+    ttl: number,
+    ifNotExists?: boolean
+  ): Promise<string | null> {
+    try {
+      if (ifNotExists) {
+        return await this.cacheService.setIfNotExist(key, JSON.stringify(val), { ttl });
+      }
+      await this.cacheService.set(key, JSON.stringify(val), { ttl });
+    } catch (err) {
+      Logger.warn(`An error occurred while setting idempotency cache, key:${key} error: ${err.message}`, LOG_CONTEXT);
+    }
+
+    return null;
+  }
+
+  private buildError(error: any): HttpException {
+    const statusCode = error.status || error.response?.statusCode || 500;
+    if (statusCode == 500 && !error.response) {
+      //some unhandled exception occured
+      return new InternalServerErrorException();
+    }
+
+    return new HttpException(error.response || error.message, statusCode, error.response?.options);
+  }
+
+  private setHeaders(response: any, headers: Record<string, string>) {
+    Object.keys(headers).map((key) => {
+      if (headers[key]) {
+        response.set(key, headers[key]);
+      }
+    });
+  }
+
+  private hashRequestBody(body: object): string {
+    const hash = createHash('blake2s256');
+    hash.update(Buffer.from(JSON.stringify(body)));
+
+    return hash.digest('hex');
+  }
+
+  private async handlerDuplicateRequest(context: ExecutionContext, bodyHash: string): Promise<Observable<any>> {
+    const cacheKey = this.getCacheKey(context);
+    const idempotencyKey = this.getIdempotencyKey(context)!;
+    const data = await this.cacheService.get(cacheKey);
+    this.setHeaders(context.switchToHttp().getResponse(), { [HEADER_KEYS.IDEMPOTENCY]: idempotencyKey });
+    const parsed = JSON.parse(data);
+    if (parsed.status === ReqStatusEnum.PROGRESS) {
+      // api call is in progress, so client need to handle this case
+      Logger.error(`previous api call in progress rejecting the request. key:${idempotencyKey}`, LOG_CONTEXT);
+      const error = 'IDEMPOTENCY_REQUEST_PROCESSING';
+      this.setHeaders(context.switchToHttp().getResponse(), {
+        [HEADER_KEYS.IDEMPOTENCY_CONFLICT]: error,
+        [HEADER_KEYS.RETRY_AFTER]: `1`,
+      });
+
+      return throwError(
+        () =>
+          new HttpException(
+            {
+              error,
+              message: `request ${idempotencyKey} is currently being processed. Please retry after 1 second`,
+            },
+            429
+          )
+      );
+    }
+    if (bodyHash !== parsed.bodyHash) {
+      //different body sent than before
+      Logger.error(`idempotency key is being reused for different bodies. key:${idempotencyKey}`, LOG_CONTEXT);
+      const error = 'IDEMPOTENCY_BODY_CONFLICT';
+      this.setHeaders(context.switchToHttp().getResponse(), { [HEADER_KEYS.IDEMPOTENCY_CONFLICT]: error });
+
+      return throwError(
+        () => new ConflictException({ error, message: `request ${idempotencyKey} is being reused for different body` })
+      );
+    }
+    //already seen the request return cached response
```

Comment:
Per https://github.com/novuhq/novu/pull/4531#issuecomment-1766557580, let's also add the `Idempotency-Replayed: true` header here when the `bodyHash` and `Idempotency-Key` match.